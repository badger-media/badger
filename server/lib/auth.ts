import NextAuth from "next-auth";
import CredentialsProvider from "next-auth/providers/credentials";
import { z } from "zod";
import { redirect } from "next/navigation";

const UserSchema = z.object({
  id: z.coerce.string(),
  first_name: z.string(),
  last_name: z.string(),
  server_name: z.string(),
  its_name: z.string(),
  email: z.string().email(),
  groups: z.array(z.string()),
});

export const {
  handlers: { GET, POST },
  auth,
  CSRF_experimental,
} = NextAuth({
  secret: process.env.AUTH_SECRET,
  providers: [
    CredentialsProvider({
      name: "YSTV Login",
      credentials: {
        username: {
          label: "Username",
          type: "text",
          placeholder: "charlie.jeffery",
        },
        password: { label: "Password", type: "password" },
      },
      async authorize(rawCredentials) {
        const credentials = z
          .object({
            username: z.string(),
            password: z.string(),
          })
          .parse(rawCredentials);
        const url = `https://sso.ystv.co.uk/REST.php?api-version=latest&api-name=usermanagement&resource-name=checkcredentials&api-username=${process.env.YSTV_SSO_USERNAME}&api-password=${process.env.YSTV_SSO_PASSWORD}`;
        const payload = new FormData();
        payload.append("username", credentials.username);
        payload.append("password", credentials.password);
        const res = await fetch(url, {
          method: "POST",
          body: payload,
        });
        if (res.status !== 200) {
          console.log(await res.text());
          return null;
        }
        return UserSchema.parse(await res.json());
      },
    }),
  ],
});

export async function requireAuth() {
  const session = await auth();
  if (!session) {
    redirect("/auth/signIn");
  }
  return session;
}
