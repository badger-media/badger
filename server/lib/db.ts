import "server-only";
import { PrismaClient } from "bowser-prisma/client";
import invariant from "tiny-invariant";

declare global {
  // eslint-disable-next-line no-var
  var prisma: PrismaClient;
}

let prisma: PrismaClient;

if (process.env.NODE_ENV === "development") {
  if (!global.prisma) {
    global.prisma = new PrismaClient({
      log: ["query", "info", "warn", "error"],
    });
  }
  prisma = global.prisma;
} else if (process.env.NODE_ENV === "test") {
  if (!global.prisma) {
    global.prisma = new PrismaClient({
      log: [
        {
          emit: "event",
          level: "query",
        },
        "info",
        "warn",
        "error",
      ],
    });
  }
  prisma = global.prisma;
  // @ts-expect-error This works, trust me.
  prisma.$on("query", async (e) => {
    // @ts-expect-error This works, trust me.
    console.log(`${e.query}\n\t${e.params}`);
  });
} else {
  invariant(process.env.DATABASE_URL, "DATABASE_URL not set");
  invariant(
    !process.env.DATABASE_URL.includes("localhost"),
    "DATABASE_URL must not be localhost in production",
  );
  prisma = new PrismaClient({
    log: ["warn", "error"],
    datasources: {
      db: {
        url: process.env.DATABASE_URL,
      },
    },
  });
}

export const db = prisma;
