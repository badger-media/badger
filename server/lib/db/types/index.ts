export * from "./show"
export * from "./rundown"
export * from "./rundownitem"
export * from "./continuityitem"
export * from "./media"
export * from "./mediaprocessingtask"
export * from "./basejob"
export * from "./processmediajob"
