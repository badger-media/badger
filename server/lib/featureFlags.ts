function flag(name: string, varName: string, def: boolean): boolean {
  return process.env[varName] === "true" || def;
}
