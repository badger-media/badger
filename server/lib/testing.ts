export const integrate =
  process.env.TEST_INTEGRATION === "true" ? describe : describe.skip;
