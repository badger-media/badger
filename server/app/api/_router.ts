import { e2eProcedure, publicProcedure, router } from "./_base";
import { z } from "zod";
import { db } from "@/lib/db";
import {
  ExtendedMediaModel,
  CompleteRundownModel,
  CompleteShowModel,
} from "@bowser/prisma/utilityTypes";
import { getPresignedURL } from "@/lib/s3";
import {
  ContinuityItemSchema,
  MediaFileSourceTypeSchema,
  MediaSchema,
  RundownItemSchema,
  ShowCreateInputSchema,
  AssetSchema,
  RundownSchema,
  ShowSchema,
  ShowUpdateInputSchema,
  MediaUpdateInputSchema,
} from "@bowser/prisma/types";
import invariant from "@/lib/invariant";
import { dispatchJobForJobrunner } from "@/lib/jobs";

const ExtendedMediaModelWithDownloadURL = ExtendedMediaModel.extend({
  continuityItems: z.array(ContinuityItemSchema),
  rundownItems: z.array(RundownItemSchema),
  assets: z.array(AssetSchema),
  downloadURL: z.string().optional().nullable(),
});

const CompleteMediaModel = ExtendedMediaModel.extend({
  continuityItems: z.array(
    ContinuityItemSchema.extend({
      show: ShowSchema,
    }),
  ),
  rundownItems: z.array(
    RundownItemSchema.extend({
      rundown: RundownSchema.extend({
        show: ShowSchema,
      }),
    }),
  ),
  assets: z.array(
    AssetSchema.extend({
      rundown: RundownSchema.extend({
        show: ShowSchema,
      }),
    }),
  ),
});

export const appRouter = router({
  ping: publicProcedure.query(() => {
    return {
      ping: "pong",
      version: process.env.NEXT_PUBLIC_VERSION!,
    };
  }),
  shows: router({
    listUpcoming: publicProcedure
      .output(z.array(ShowSchema))
      .query(async () => {
        return db.showWithDuration.findMany({
          where: {
            end: {
              gte: new Date(),
            },
          },
        });
      }),
    get: publicProcedure
      .input(z.object({ id: z.number() }))
      .output(CompleteShowModel)
      .query(async ({ input }) => {
        return db.show.findUniqueOrThrow({
          where: {
            id: input.id,
          },
          include: {
            continuityItems: {
              include: {
                media: true,
              },
              orderBy: {
                order: "asc",
              },
            },
            rundowns: {
              include: {
                items: {
                  include: {
                    media: true,
                  },
                  orderBy: {
                    order: "asc",
                  },
                },
                assets: {
                  include: {
                    media: true,
                  },
                },
              },
            },
          },
        });
      }),
    getVersion: publicProcedure
      .input(z.object({ id: z.number() }))
      .output(z.object({ version: z.number() }))
      .query(async ({ input }) => {
        const obj = await db.show.findUniqueOrThrow({
          where: {
            id: input.id,
          },
          select: {
            version: true,
          },
        });
        return obj;
      }),
    create: e2eProcedure
      .input(ShowCreateInputSchema)
      .output(CompleteShowModel)
      .mutation(async ({ input }) => {
        const res = await db.show.create({
          data: input,
          include: {
            continuityItems: {
              include: {
                media: true,
              },
            },
            rundowns: {
              include: {
                items: {
                  include: {
                    media: true,
                  },
                },
                assets: {
                  include: {
                    media: true,
                  },
                },
              },
            },
          },
        });
        return res;
      }),
    update: e2eProcedure
      .input(
        z.object({
          id: z.number(),
          data: ShowUpdateInputSchema,
        }),
      )
      .output(CompleteShowModel)
      .mutation(async ({ input }) => {
        const res = await db.show.update({
          where: {
            id: input.id,
          },
          data: input.data,
          include: {
            continuityItems: {
              include: {
                media: true,
              },
            },
            rundowns: {
              include: {
                items: {
                  include: {
                    media: true,
                  },
                },
                assets: {
                  include: {
                    media: true,
                  },
                },
              },
            },
          },
        });
        return res;
      }),
  }),
  media: router({
    get: publicProcedure
      .input(z.object({ id: z.number() }))
      .output(ExtendedMediaModelWithDownloadURL)
      .query(async ({ input }) => {
        // Need this Omit to ensure that this findFirstOrThrow works while assuring type
        // safety for all the other fields of ExtendedMediaModelWithDownloadURL
        const obj: Omit<
          z.infer<typeof ExtendedMediaModelWithDownloadURL>,
          "downloadURL"
        > = await db.media.findFirstOrThrow({
          where: {
            id: input.id,
          },
          include: {
            tasks: true,
            continuityItems: true,
            rundownItems: true,
            assets: true,
          },
        });
        if (obj.path !== null) {
          (
            obj as z.infer<typeof ExtendedMediaModelWithDownloadURL>
          ).downloadURL = await getPresignedURL(obj.path);
        }
        return obj as z.infer<typeof ExtendedMediaModelWithDownloadURL>;
      }),
    create: e2eProcedure
      .input(
        z.object({
          sourceType: MediaFileSourceTypeSchema,
          source: z.string(),
          fileName: z.string(),
          targetType: z.enum(["rundownItem", "continuityItem"]), // TODO: support assets
          targetID: z.number(),
          process: z.boolean(),
        }),
      )
      .output(MediaSchema)
      .mutation(async ({ input }) => {
        const [media, job] = await db.$transaction(async ($db) => {
          let med;
          switch (input.targetType) {
            case "rundownItem":
              med = await $db.media.create({
                data: {
                  name: input.fileName,
                  rawPath: "",
                  durationSeconds: 0,
                  rundownItems: {
                    connect: {
                      id: input.targetID,
                    },
                  },
                },
              });
              await $db.rundownItem.update({
                where: {
                  id: input.targetID,
                },
                data: {
                  rundown: {
                    update: {
                      show: {
                        update: {
                          version: {
                            increment: 1,
                          },
                        },
                      },
                    },
                  },
                },
              });
              break;
            case "continuityItem":
              med = await $db.media.create({
                data: {
                  name: input.fileName,
                  rawPath: "",
                  durationSeconds: 0,
                  continuityItems: {
                    connect: {
                      id: input.targetID,
                    },
                  },
                },
              });
              await $db.continuityItem.update({
                where: {
                  id: input.targetID,
                },
                data: {
                  show: {
                    update: {
                      version: {
                        increment: 1,
                      },
                    },
                  },
                },
              });
              break;
            default:
              invariant(false, "Invalid target type " + input.targetType);
          }
          let job;
          if (input.process) {
            job = await $db.processMediaJob.create({
              data: {
                sourceType: input.sourceType,
                source: input.source,
                media: {
                  connect: {
                    id: med.id,
                  },
                },
                base_job: {
                  create: {},
                },
              },
            });
          }
          return [med, job] as const;
        });
        if (job) {
          await dispatchJobForJobrunner(job.base_job_id);
        }
        return media;
      }),
    update: e2eProcedure
      .input(
        z.object({
          id: z.number(),
          data: MediaUpdateInputSchema,
        }),
      )
      .output(MediaSchema)
      .mutation(async ({ input }) => {
        return db.media.update({
          where: {
            id: input.id,
          },
          data: input.data,
        });
      }),
    bulkGet: publicProcedure
      .input(z.array(z.number()))
      .output(z.array(CompleteMediaModel))
      .query(async ({ input }) => {
        return await db.media.findMany({
          where: {
            id: {
              in: input,
            },
          },
          include: {
            tasks: true,
            continuityItems: {
              include: {
                show: true,
              },
            },
            rundownItems: {
              include: {
                rundown: {
                  include: {
                    show: true,
                  },
                },
              },
            },
            assets: {
              include: {
                rundown: {
                  include: {
                    show: true,
                  },
                },
              },
            },
          },
        });
      }),
  }),
  rundowns: router({
    get: publicProcedure
      .input(z.object({ id: z.number() }))
      .output(CompleteRundownModel)
      .query(async ({ input }) => {
        return db.rundown.findUniqueOrThrow({
          where: {
            id: input.id,
          },
          include: {
            items: {
              include: {
                media: true,
              },
            },
            assets: {
              include: {
                media: true,
              },
            },
          },
        });
      }),
  }),
});

export type AppRouter = typeof appRouter;
