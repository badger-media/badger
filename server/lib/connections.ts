import { ConnectionTarget } from "@bowser/prisma/client";
import { db } from "./db";
import { cookies } from "next/headers";
import { OAuth2Client } from "google-auth-library";
import { expectNever } from "ts-expect";

export function makeGoogleOauthClient() {
  return new OAuth2Client({
    clientId: process.env.GOOGLE_CLIENT_ID!,
    clientSecret: process.env.GOOGLE_CLIENT_SECRET!,
    redirectUri: `${process.env.PUBLIC_URL!}/connect/google/callback`,
  });
}

function connCookieName(target: ConnectionTarget) {
  return `${target}_access_token`;
}

function getConnCookie(target: ConnectionTarget) {
  return cookies().get(connCookieName(target));
}

function setConnCookie(target: ConnectionTarget, value: string) {
  return cookies().set(connCookieName(target), value, {
    sameSite: "strict",
    expires: new Date(Date.now() + 60 * 60 * 1000),
    secure: process.env.NODE_ENV === "production",
    httpOnly: true,
  });
}

export async function getConnectionAccessToken(
  target: ConnectionTarget,
  userID: number,
) {
  const cookie = getConnCookie(target);
  if (cookie) {
    return cookie.value;
  }

  const conn = await db.connection.findUnique({
    where: {
      userId_target: {
        target,
        userId: userID,
      },
    },
  });
  if (!conn) {
    return null;
  }

  switch (target) {
    case ConnectionTarget.google: {
      const client = makeGoogleOauthClient();
      client.setCredentials({
        refresh_token: conn.refreshToken,
      });
      const accessToken = await client.getAccessToken();
      setConnCookie(target, accessToken.token!);
      return accessToken.token;
    }
    default:
      expectNever(target);
  }
}

export async function saveTokens(
  target: ConnectionTarget,
  userID: number,
  accessToken: string,
  refreshToken?: string | null,
) {
  setConnCookie(target, accessToken);
  if (refreshToken) {
    await db.connection.upsert({
      where: {
        userId_target: {
          userId: userID,
          target,
        },
      },
      update: {
        refreshToken: refreshToken,
      },
      create: {
        target,
        userId: userID,
        refreshToken: refreshToken,
      },
    });
  }
}
