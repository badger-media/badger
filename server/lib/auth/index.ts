import { makeJWT, parseAndVerifyJWT } from "@/lib/auth/jwt";
import { NextRequest } from "next/server";
import { redirect } from "next/navigation";
import * as Sentry from "@sentry/nextjs";
import { Permission } from "@badger/prisma/client";
import { db } from "../db";
import { User, UserSchema } from "@badger/prisma/types";
import {
  autoActivateAllUsers,
  disablePermissionsChecks,
  enableUserManagement,
} from "@badger/feature-flags";
import { BasicUserInfo } from "./types";
import { Forbidden, Unauthorized } from "./errors";

// HACK: middleware runs in the edge runtime and for some reason fails the server-only check
if (process.env.NEXT_RUNTIME !== "edge") {
  // @ts-expect-error idk lol
  import("server-only");
}

export enum SignInResult {
  Success = "success",
  CreatedInactive = "created_inactive",
  Inactive = "inactive",
}

const cookieName = "badger_session";

export async function doSignIn(
  provider: string,
  credentials: BasicUserInfo,
): Promise<SignInResult> {
  let user: User;
  if (enableUserManagement) {
    const autoActivateDomains = new Set(
      process.env.USER_AUTO_CREATE_DOMAINS?.split(", "),
    );
    const shouldAutoActivate =
      autoActivateAllUsers ||
      autoActivateDomains.has(credentials.domain ?? "NEVER");
    let didExist;
    [user, didExist] = await db.$transaction(async ($db) => {
      const user = await $db.user.findFirst({
        where: {
          identities: {
            some: {
              provider,
              identityID: credentials.id,
            },
          },
        },
      });
      if (user) {
        return [user, true] as const;
      }
      return [
        await $db.user.create({
          data: {
            name: credentials.name,
            email: credentials.email ?? null,
            permissions: shouldAutoActivate ? [Permission.Basic] : [],
            isActive: shouldAutoActivate,
            identities: {
              create: {
                provider,
                identityID: credentials.id,
              },
            },
          },
        }),
        false,
      ] as const;
    });
    if (!didExist && !shouldAutoActivate) {
      return SignInResult.CreatedInactive;
    }
    if (!user.isActive) {
      return SignInResult.Inactive;
    }
  } else {
    user = {
      id: parseInt(credentials.id, 10),
      email: credentials.email ?? null,
      isActive: true,
      name: credentials.name,
      permissions: [Permission.Basic],
    };
  }

  const claims = user as Record<string, unknown>;
  claims.v = 2;
  claims.iss = process.env.PUBLIC_URL;
  claims.sub = user.id;
  const iat = Math.floor(Date.now() / 1000);
  claims.iat = iat;
  claims.nbf = iat;
  claims.exp = iat + 60 * 60 * 24 * 7;
  const token = await makeJWT(claims);
  const { cookies } = await import("next/headers");
  cookies().set(cookieName, token, {
    maxAge: 60 * 60 * 24 * 7,
  });
  if (Sentry.getCurrentHub().getClient()) {
    Sentry.setUser({
      id: user.id,
      email: user.email ?? undefined,
    });
  }

  return SignInResult.Success;
}

export function makePublicURL(baseUrl: URL | string) {
  const url = new URL(baseUrl.toString(), process.env.PUBLIC_URL);
  // Just passing the second parameter to `new URL()` isn't sufficient to *override* the host
  if (process.env.PUBLIC_URL) {
    const publicUrl = new URL(process.env.PUBLIC_URL);
    url.protocol = publicUrl.protocol;
    url.host = publicUrl.host;
    url.port = publicUrl.port;
  }
  return url.toString();
}

async function getSessionFromCookie(req?: NextRequest) {
  if (req) {
    const sessionID = req.cookies.get(cookieName);
    if (!sessionID) return null;
    return sessionID.value;
  }
  const { cookies } = await import("next/headers");
  const sessionID = cookies().get(cookieName);
  if (!sessionID) return null;
  return sessionID.value;
}

export async function deleteSession() {
  const { cookies } = await import("next/headers");
  cookies().delete(cookieName);
}

export async function checkSession(req?: NextRequest) {
  const sid = await getSessionFromCookie(req);
  if (!sid) return null;
  let rawUser;
  try {
    rawUser = await parseAndVerifyJWT(sid, {
      iss: process.env.PUBLIC_URL!,
    });
  } catch (e) {
    console.warn("Failed to parse JWT", e);
    return null;
  }
  if (rawUser.v !== 2) {
    console.warn("Got outdated session JWT, treating as unauthenticated");
    return null;
  }
  const user = UserSchema.parse(rawUser);
  if (Sentry.getCurrentHub().getClient()) {
    Sentry.setUser({
      id: user.id,
      email: user.email ?? undefined,
    });
  }
  if (!user.isActive) {
    // TODO[BOW-108]: This doesn't check it from the database, meaning that if the user is deactivated
    //  while signed in, they can still access the site until their session expires.
    return null;
  }
  return user;
}

export async function auth(req?: NextRequest) {
  const u = await checkSession(req);
  if (!u) {
    redirect(
      makePublicURL("/login?return=" + makePublicURL(req?.nextUrl ?? "/")),
    );
  }
  return u;
}

export async function hasPermission(perm: Permission, req?: NextRequest) {
  const user = await checkSession(req);
  if (!user) return false;
  if (disablePermissionsChecks) return true;
  if (user.permissions.includes(perm)) return true;
  if (user.permissions.includes(Permission.SUDO)) return true;
  return false;
}

export async function requirePermission(perm: Permission, req?: NextRequest) {
  const user = await checkSession(req);
  if (!user) throw new Unauthorized();
  if (disablePermissionsChecks) return user;
  if (user.permissions.includes(perm)) return user;
  if (user.permissions.includes(Permission.SUDO)) return user;
  throw new Forbidden(perm);
}
