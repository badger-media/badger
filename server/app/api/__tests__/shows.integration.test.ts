import { integrate } from "@badger/testing";
import { appRouter } from "../_router";
import { db } from "@/lib/db";
import { add, sub } from "date-fns";

const api = appRouter.createCaller({});

integrate("shows", () => {
  beforeEach(async () => {
    await db.$executeRawUnsafe("DELETE FROM shows");
  });
  describe("listUpcoming", () => {
    it("returns nothing with no shows", async () => {
      await expect(api.shows.listUpcoming()).resolves.toEqual([]);
    });

    it("returns upcoming shows", async () => {
      await db.show.create({
        data: {
          name: "Test Show",
          start: add(new Date(), { hours: 1 }),
        },
      });
      const r = await api.shows.listUpcoming();
      expect(r).toHaveLength(1);
      expect(r[0]).toMatchSnapshot({
        id: expect.any(Number),
        name: "Test Show",
        start: expect.any(Date),
        version: 0,
      });
    });

    it("excludes shows that have already finished", async () => {
      await db.show.create({
        data: {
          name: "Test Show",
          start: sub(new Date(), { hours: 1 }),
        },
      });
      await expect(api.shows.listUpcoming()).resolves.toEqual([]);
    });

    it("returns shows that have started but not finished [BDGR-112]", async () => {
      await db.show.create({
        data: {
          name: "Test Show",
          start: sub(new Date(), { minutes: 15 }),
          continuityItems: {
            create: {
              name: "Test",
              order: 0,
              durationSeconds: 30 * 60,
            },
          },
        },
      });
      const r = await api.shows.listUpcoming();
      expect(r).toHaveLength(1);
    });
  });
});
