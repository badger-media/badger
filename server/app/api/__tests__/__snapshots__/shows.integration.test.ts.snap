// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows listUpcoming returns upcoming shows 1`] = `
{
  "id": Any<Number>,
  "name": "Test Show",
  "start": Any<Date>,
  "version": 0,
}
`;
